Local deploy using env files

sergioifg94 opened the pull request:
Closes #19 

Add a kustomization that creates a ConfigMap and Secret from the local env files and mounts them in the controller Pod

## Verification steps

> Ensure you have the correct values in the local env files `controller-config.env` and `aws-credentials.env`

1. Run local setup
    ```sh
    make local-setup
    ```
2. Build and load the image
    ```sh
    make docker-build
    kind load docker-image controller:latest --name mctc-control-plane  --nodes mctc-control-plane-control-plane    
    ```
3. Deploy the controller
    ```sh
    make deploy
    ```
4. Verify that the controller container runs correctly

mikenairn commented on Makefile:
```diff
@@ -121,7 +121,7 @@ uninstall: manifests kustomize ## Uninstall CRDs from the K8s cluster specified
 deploy: manifests kustomize ## Deploy controller to the K8s cluster specified in ~/.kube/config.
 	cd config/manager && $(KUSTOMIZE) edit set image controller=${IMG}
-	$(KUSTOMIZE) build config/default | kubectl apply -f -
+	$(KUSTOMIZE) --load-restrictor LoadRestrictionsNone build config/deploy/local | kubectl apply -f -
```
What does `--load-restrictor LoadRestrictionsNone` do?

sergioifg94 replied:
It disables the restriction that the env files have to be under the kustomization directory, so we can reference them from the root of the project

mikenairn replied:
Interesting, good to know.

mikenairn commented on config/deploy/local/kustomization.yaml:
```diff
@@ -0,0 +1,27 @@
+secretGenerator:
+  - name: aws-credentials
+    envs:
+      - ../../../aws-credentials.env
```
I think in kcp-glbc we just had these always be generated for all deployments and merged in local values. Do we see any value in just doing that here, since presumably we need to inject these config/credential values for all deployments anyway, not just loical.

mikenairn replied:
Presumably these files need to exist, does anything make sure of that, or does the deploy just fail and we have it in the readme or soemthing?

mikenairn commented:
This is fine for local dev. Although i don't think it fully resolves what was being asked for in https://github.com/Kuadrant/multi-cluster-traffic-controller/issues/19. Even though that does mention `make deploy`, the solution we used for kcp-glbc would work for all deployments and not just local deployments which this does.

Do you see a deployment kustomize overlay having to inject all of this in this way? Is there ever a time when a deployment is not going to need an aws credentials secret?

sergioifg94 commented:
I updated the PR to have the mounting of the environment variables by default in `config/manager/manager.yaml`, and I updated the README to reflect that the deployment simply needs to have the env files instead of modifying the manifest.

I didn't add the functionallity to merge environment variables for the sake of simplicity as it might become confusing to know where what values come from

mikenairn reviewed (approved):
Looks fine. Small query, but nothing that needs to hold up a merge.

sergioifg94 commented on config/deploy/local/manager_config_patch.yaml:
```diff
@@ -0,0 +1,11 @@
+    spec:
+      containers:
+      - name: manager
+        imagePullPolicy: Never
```
Yes. I assume that the local deploy will be built for the `make deploy` target, and everything will run in the kind clusters. If there's a reason to assume otherwise I'm happy to change it

mikenairn commented on README.md:
````diff
@@ -31,25 +31,9 @@ When deploying the multi cluster traffic controller using the make targets the f
-    ```
+1. Create two env files:
+    * One called `aws-credentials.env` containing **AWS_ACCESS_KEY_ID** and **AWS_SECRET_ACCESS_KEY**
+    * One called `controller-config` containing **AWS_DNS_PUBLIC_ZONE_ID** and **ZONE_ROOT_DOMAIN**
````
Just realised this should be `controller-config.env`. Anyway, can be updated in another PR.
